shipment_planning_maxoptra/models/res_partner.py

```diff
@@ -15,6 +16,26 @@ class ResPartner(models.Model):
         "partner to set on Batch pickings after import.",
     )
 
+    maxoptra_partner_key = fields.Text(string="Maxoptra Key")
+
+    @api.constrains("maxoptra_partner_key")
+    def _check_maxoptra_partner_key(self):
+        """ Check that the maxoptra_partner_key is unique """
+        for record in self:
+            existing_partners_ids = (
+                self.env["res.partner"].search(
+                    [("maxoptra_partner_key", "=", record.maxoptra_partner_key)]
+                )
+                - record
+            )
+            if existing_partners_ids:
```

Comment:
```suggestion
            if not record.maxoptra_partner_key:
                continue
            same_maxoptra_key_exists = (
                self.env["res.partner"].search_count(
                    [("maxoptra_partner_key", "=", record.maxoptra_partner_key), ("id", "!=", record.id)]
                )
            )
            if same_maxoptra_key_exists:
```
Few things that will improve performance:
 - Avoid the search if maxoptra_partner_key is undefined
 - Use a domain condition instead of `-` operation on recordset
 - Use `search_count` to avoid browsing the records

Reply:
PS: to nitpick, I would call the field `maxoptra_key` instead of `maxoptra_partner_key` since it's defined on `res.partner` it must be for a partner anyway :wink:

Reply:
let's leave this field name as it is for now, already renamed in once before weekend.